AI vision overhauls.

TsunamiAnt opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

I had the majority of this done already so thought fuck it why not finish it. I scrapped a few things I was going to do in here and am now just focusing on the camera and vision improvements.

- AI's can now interact with objects even while they are in camera static by use of a small hologram indicator.
- These can be disabled by cutting the AI wire on the area APC.
- AI tracking now functions mainly off of suit sensors and even allows tracking in obscured areas, provided the sensors are on, of course. This will take 2 seconds. Facial recognition is now impossible if the face is obscured, and generally is likely to take longer.
- Any camera in the AIs vision(similar to camera lights) now shows a red blip no matter what. The AI cannot stop this from happening. If the AI is looking, you will know it.
- Camera vision reduction with multitools has been lowered to 1 tile.

Now, I did revive a 5 month old branch for this, so if there are changes in here I haven't documented, that's entirely because I myself did not know they were in here. I'll get to review my own PR! How fun.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

It makes no dang sense that the AI needs visuals to interact with networked machinery.
Simultaneuously(?) nerfs and buffs cameras in such a way as to make disabling them more interesting with more concrete counterplay, yet also ease up some qol for the AI.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

morty, the camera is RED, morty!!
<img width="463" height="358" alt="image" src="https://github.com/user-attachments/assets/07656f4d-4e0f-4344-9430-51116aac9160" />

Let it be light
<img width="764" height="852" alt="image" src="https://github.com/user-attachments/assets/72c19249-3a71-4f18-b264-94812a6a2b71" />

Let it be dark
<img width="821" height="782" alt="image" src="https://github.com/user-attachments/assets/0af02056-7a76-4a5e-805a-8c055bbb3429" />

oooh we cuttin'
<img width="719" height="682" alt="image" src="https://github.com/user-attachments/assets/0da4e014-de76-4b2c-aaf1-93c4ed987123" />

with it cut, AI can't see SHIET
<img width="701" height="808" alt="image" src="https://github.com/user-attachments/assets/8011d953-2279-47da-9e7f-6a71870f8a4f" />

</details>

## Changelog
:cl:
add: Added little holograms for AIs that they can see over camera static that lets them interact with machines :)
add: Added the ability for the AI wire on APCs to disable that :(
tweak: tweaked Camera short viewrange smaller
qol: made camera tracking more robust with active suit sensor beacon, less robust without.
balance: If an AI is seeing you through a camera, the camera will have a red lens.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TsunamiAnt committed:
first one

TsunamiAnt committed:
weee

TsunamiAnt committed:
wah

DrDuckedGoose committed:
Plane Master & Effects - 18 7 25

TsunamiAnt committed:
Merge pull request #3 from DrDuckedGoose/AIUploadRestrictions-Racc

Plane Master & Effects - 18 7 25

TsunamiAnt committed:
Attempt 1

TsunamiAnt committed:
Merge remote-tracking branch 'BeeStation/master' into AIUploadRestrictions

TsunamiAnt committed:
Adds AI view for specific atoms

Adds a dedicated AI view to certain atoms, allowing them to be visible through fog of war and other visual obstructions.
This involves adding new variables and procs to manage the AI appearance and its visibility, improving AI's situational awareness.

TsunamiAnt committed:
Improves AI view handling and restrictions.

Adds checks for `ai_view` atoms to be interactable even outside camera view.

Moves hologram appearance pixel offsets to be set directly.

Sets `ai_view` to false for portable atmospherics and power terminals by default. This restricts AI camera usage on objects.

TsunamiAnt committed:
Updates AI views on APC wire manipulation

Ensures AI view overlays are correctly updated when APC AI wires are cut or mended.

This change provides visual feedback to the AI about the status of areas affected by APC wiring changes. It calls a new proc to update AI view status within the area of the APC.

TsunamiAnt committed:
Reduces camera view range

Reduces the default camera view range to improve performance and balance gameplay.

The original value was excessively large and unnecessary.

TsunamiAnt committed:
Hides machinery from AI view

Adds ai_view = FALSE to several machinery types.

This prevents the AI from directly interacting with or being influenced by these machines, which improves gameplay balance and prevents unintended AI behaviors.

TsunamiAnt committed:
Refactors AI tracking logic for humans

Overhauls the AI's target tracking system, prioritizing suit sensor data for humans.

Improves tracking efficiency when suit sensors are available, reducing tracking time.
Tweaks facial recognition tracking and removes reliance on camera if the suit sensors are transmitting.
Adjusts tracking times based on sensor availability, lighting, and target type.
Updates relevant chat messages to reflect changes to tracking mechanics.

TsunamiAnt committed:
Adds camera telegraphing to AI vision.

Implements a system for the AI to telegraph which cameras it is viewing,
displaying an "in use" indicator on those cameras.

This is done to improve gameplay feedback when an AI is watching a camera feed.

Additionally, this change ensures proper cleanup of these indicators when
the AI is destroyed or its vision is disrupted, preventing orphaned indicators.

TsunamiAnt committed:
Increases camera view range.

Increases the camera's default view range.

This change makes it slightly easier to monitor larger areas with cameras.

TsunamiAnt committed:
Merge branch 'master' into AIUploadRestrictions

TsunamiAnt committed:
Adds cut AI wire mapping helper

Adds the cut AI wire mapping helper to several maps in order to facilitate the placement of cut wires in APCs.

Increases camera view range to improve AI visibility.

TsunamiAnt committed:
Merge remote-tracking branch 'origin/AIUploadRestrictions' into AIUploadRestrictions

TsunamiAnt committed:
Merge remote-tracking branch 'BeeStation/master' into AIUploadRestrictions

TsunamiAnt committed:
Hides machinery from AI view

Hides certain machinery types from AI view to prevent unwanted AI interactions or behavior.

Specifically, this change hides:
- Mineral stacking unit consoles
- Conveyor belts
- Disposal bins

TsunamiAnt committed:
Merge remote-tracking branch 'BeeStation/master' into AIUploadRestrictions

PowerfulBacon reviewed (changes requested):
This looks really good, but something needs to be done about the ease of spotting things that are out of place. One of the largest problems with AI is the helicopter AI who can ruin a round extremely quickly by spotting one thing out of place. The two key issues here are:
- The AI can see and interact with machines built out of their view
- The player has no obvious indicator that the AI can see any machine

My primary suggestion would be to make it so that the AI cannot see what machines are inside of maintenance by default.

Second possible suggestion would be to make it so that all computers have a button somewhere on the UI (maybe in the window) that let you toggle the AI connection, though this would be challenging as it would require sending ui_data for every single computer which won't share a common ui_data parent.

You can also track people that you can't see, which includes being able to follow them wherever they go. I don't like being able to freely track players who aren't on cameras, even if they have suit sensors on since sometimes you just forget to turn them off. I'm not entirely against this being a feature, but it needs to have an indication to the player about when they are being tracked. If the AI can track people without having any cameras nearby then it negates the benefit of cameras showing when the AI is watching you (since now the AI can watch you without even having a camera at your location).

## Requirements

- Antagonists must be able to construct a machine in maintenance stealthily without the possibility that the AI stumbles across what was constructed, and the method in which you prevent the AI from spoiling your round in this way must be obvious to the player (simply telling players to cut the APC wire is not enough when this is a system hidden in the background which isn't presented to players in any way).
- Either limit the off-camera tracking, or give an obvious indicator to someone as to when they are being tracked by the AI. The worst type of AI unfairness is when they find something out about you without you realizing and then stealthily relay this information to security without you having any way of verifying that.
- All machines and trackability must respond to the signal jammer, if a signal jammer is active then machines in its radius should not be visible to the AI. (Similar to how cameras are disabled if a signal jammer is activated near them)

## Preferences

- Limit the amount of time that someone is trackable when they exit suit sensor range, so someone can be tracked for (say) 10 seconds with suit sensors maxed out, but after that the track is lost and it no longer functions.
- Malf AI needs the ability to forcibly re-enable an APCs connection. Disabling AI interaction in huge and important areas will be extremely easy.
- Malf AI hacked APCs should never be able to be disconnected from the network and should be forcibly enabled

Requirements are essentially reducing the possibility that the AI can further ruin rounds early by finding information in a way that is hidden/unfair to the player.

To clarify a bit more of where I stand, this is my opinion of each individual change:

- AI's can now interact with objects even while they are in camera static by use of a small hologram indicator. **(Positive for visuals, negative for implementation in how nowhere is safe from the AI which generally is heavily security-sided)**
- These can be disabled by cutting the AI wire on the area APC. **(Very Positive)**
- AI tracking now functions mainly off of suit sensors and even allows tracking in obscured areas, provided the sensors are on, of course. This will take 2 seconds. Facial recognition is now impossible if the face is obscured, and generally is likely to take longer.  **(Very Negative)**
- Any camera in the AIs vision(similar to camera lights) now shows a red blip no matter what. The AI cannot stop this from happening. If the AI is looking, you will know it. **(Extremely Positive, this change has been a long time coming)**
- Camera vision reduction with multitools has been lowered to 1 tile.**(Positive, minimal impact)**

TsunamiAnt committed:
Restricts AI uploads to visible APCs.

Restricts AI uploads to APCs that are either directly connected to the AI, or are visible and hackable.

Adds a system to 'hack' APCs remotely if they are visible but not directly connected.
This allows the AI to bypass the 'no direct connection' restriction.

This change also refactors AI control logic for APCs into a single proc.

TsunamiAnt committed:
Merge branch 'master' into AIUploadRestrictions

TsunamiAnt committed:
Adds AI hacking indicators to APCs

Adds visual and audible cues to APCs when an AI is attempting to hack them.

This change enhances player awareness of AI intrusion attempts and provides
players with a method of stopping the breach. Prevents AI hacking attempts
from succeeding when the APC is not within camera range.

TsunamiAnt committed:
Adds tracking telegraph functionality

Adds a system where tracked targets with enhanced sensors will trigger a "ping" sound effect and visual cue if not on camera.

This provides feedback to the AI that their target is aware of being tracked and may take countermeasures.

Improves the AI's gameplay experience by giving more information about the target's awareness.

TsunamiAnt committed:
I do not get this logic, does this work?
